casper/src/test/scala/coop/rchain/casper/addblock/ProposerSpec.scala

```diff
@@ -0,0 +1,227 @@
+package coop.rchain.casper.addblock
+
+import cats.Applicative
+import cats.effect.Resource
+import cats.effect.concurrent.Deferred
+import cats.syntax.all._
+import coop.rchain.casper._
+import coop.rchain.casper.blocks.proposer._
+import coop.rchain.casper.helper.{BlockDagStorageFixture, NoOpsCasperEffect}
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.casper.util.GenesisBuilder.defaultValidatorSks
+import coop.rchain.casper.util.rholang.Resources.mkRuntimeManager
+import coop.rchain.casper.util.rholang.{Resources, RuntimeManager}
+import coop.rchain.metrics.Metrics.MetricsNOP
+import coop.rchain.metrics.{NoopSpan, Span}
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.models.blockImplicits.getRandomBlock
+import coop.rchain.shared.Log
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import org.scalatest.{FlatSpec, Matchers}
+
+import scala.collection.immutable.HashMap
+
+class ProposerSpec extends FlatSpec with Matchers with BlockDagStorageFixture {
```

Comment:
Hm.....I don't know how to put this. The tests here seem to test themselves instead of testing something very important in production. I used to write tests like these in other projects.

Reply:
Oh. Nvm. I missed something. They are actually good.